### ISSUE TITLE
Maven plugin test-update goal is not working

### ISSUE THREAD

**Issue by blofroth** [Apicurio/apicurio-registry#1248]
Testing running a non-compatible schema with "test-update" does not fail the maven build.

It seems nothing special is done when the result map is populated with failed entries:
https://github.com/Apicurio/apicurio-registry/blob/master/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/TestUpdateRegistryMojo.java#L65

Tested schemas was uploading this:
https://github.com/confluentinc/examples/blob/6.1.0-post/clients/avro/src/main/resources/avro/io/confluent/examples/clients/basicavro/Payment2a.avsc

Where this was the existing schema:
https://github.com/confluentinc/examples/blob/6.1.0-post/clients/avro/src/main/resources/avro/io/confluent/examples/clients/basicavro/Payment.avsc

Only on non "bad requests" (errors) does it fail the build:
https://github.com/Apicurio/apicurio-registry/blob/master/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/TestUpdateRegistryMojo.java#L83

Probably there needs to be some check for the invalid entries:
```
List<Map.Entry<String, Boolean>> invalidArtifacts =
        results.entrySet().stream().filter(e -> Boolean.FALSE.equals(e.getValue()))
            .collect(Collectors.toList());

if(!invalidArtifacts.isEmpty()) { throw ...}
```

**Comment by EricWittmann** [comment]
I assume you enabled the Compatibility rule either globally or for the artifact?  The Test Update operation will call the appropriate server-side API endpoint, but if no Rules are enabled, then it won't do anything.

Just wanted to check that before diving into the problem. :)

**Comment by blofroth** [comment]
Yes, I could see HTTP 400 in the access log for the schema registry Ingress. 

I also tested to upload the same artifact in the GUI, where it was rejected.

The only difference for a succesful "test-update" run is this output line:
https://github.com/Apicurio/apicurio-registry/blob/master/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/TestUpdateRegistryMojo.java#L61

But both variants succeed as Maven executions without error.

**Comment by EricWittmann** [comment]
OK thanks for the details.  Clearly a bug then.  Happy to have a contribution if you wanted to try a fix.  If not that's cool too, we'll get this fixed for 2.0.

**Comment by blofroth** [comment]
I can probably take a stab at it.

I tried to replicate locally with latest master, but couldn't. So somethings going on, hopefully it was not a silly mistake on my part with my first test. I will try to replicate that scenario again and post more details.

I also experienced another issue with Confluent Json Schema serializers. Hopefully I can contribute a replicating test case for that and file another ticket.

**Comment by blofroth** [comment]
Just checking I wasn't crazy, I could reproduce it on a branch based off 1.3.2.Final:
https://github.com/Apicurio/apicurio-registry/compare/1.3.2.Final...blofroth:reproduce-1248

Starting the registry with:
```
cd apicurio-registry/app
mvn compile quarkus:dev -Dquarkus.http.access-log.enabled=true
```

Uploading a schema and then test-update for incompatible schema:
```
cd utils/maven-plugin/src/test/resources/test-build
# register
mvn install
# manually enable compat rule in UI
# test update
test-build % mvn  install -Dschema.path=src/main/resources/schemas/recordIncompatible.avsc -Ptest-update 
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------< io.apicurio:apicurio-test-maven-plugin >---------------
[INFO] Building apicurio-test-maven-plugin 1.0-SNAPSHOT
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- apicurio-registry-maven-plugin:1.3.2.Final:test-update (default) @ apicurio-test-maven-plugin ---
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by retrofit2.Platform (file:/Users/blofroth/.m2/repository/com/squareup/retrofit2/retrofit/2.9.0/retrofit-2.9.0.jar) to constructor java.lang.invoke.MethodHandles$Lookup(java.lang.Class,int)
WARNING: Please consider reporting this to the maintainers of retrofit2.Platform
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ apicurio-test-maven-plugin ---
[INFO] Installing <path>/apicurio-registry/utils/maven-plugin/src/test/resources/test-build/pom.xml to /Users/blofroth/.m2/repository/io/apicurio/apicurio-test-maven-plugin/1.0-SNAPSHOT/apicurio-test-maven-plugin-1.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.184 s
[INFO] Finished at: 2021-02-27T18:42:11+01:00
[INFO] ------------------------------------------------------------------------
```

Quarkus logs says:
```
2021-02-27 18:42:11,071 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) RESTEASY002315: PathInfo: /artifacts/user-schema/test
2021-02-27 18:42:11,072 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) Interceptor Context: org.jboss.resteasy.core.interception.jaxrs.ServerReaderInterceptorContext,  Method : proceed
2021-02-27 18:42:11,072 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyReader: org.jboss.resteasy.core.providerfactory.SortedKey
2021-02-27 18:42:11,072 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyReader: org.jboss.resteasy.plugins.providers.InputStreamProvider
2021-02-27 18:42:11,072 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) Provider : org.jboss.resteasy.plugins.providers.InputStreamProvider,  Method : readFrom
2021-02-27 18:42:11,072 DEBUG [io.api.reg.res.ArtifactsResourceImpl_Subclass] (executor-thread-3) ENTERING method [testUpdateArtifact] with 3 parameters
2021-02-27 18:42:11,073 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) ENTERING method [getArtifactRules] with 1 parameters
2021-02-27 18:42:11,074 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) LEAVING method [getArtifactRules]
2021-02-27 18:42:11,074 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) ENTERING method [getArtifact] with 1 parameters
2021-02-27 18:42:11,074 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) LEAVING method [getArtifact]
2021-02-27 18:42:11,074 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) ENTERING method [getArtifactRule] with 2 parameters
2021-02-27 18:42:11,075 DEBUG [io.api.reg.sto.imp.InMemoryRegistryStorage_Subclass] (executor-thread-3) LEAVING method [getArtifactRule]
2021-02-27 18:42:11,075 DEBUG [io.api.reg.rul.com.CompatibilityRuleExecutor_Subclass] (executor-thread-3) ENTERING method [execute] with 1 parameters
2021-02-27 18:42:11,075 DEBUG [io.api.reg.typ.pro.ArtifactTypeUtilProviderImpl_Subclass] (executor-thread-3) ENTERING method [getArtifactTypeProvider] with 1 parameters
2021-02-27 18:42:11,075 DEBUG [io.api.reg.typ.pro.ArtifactTypeUtilProviderImpl_Subclass] (executor-thread-3) LEAVING method [getArtifactTypeProvider]
2021-02-27 18:42:11,075 DEBUG [io.api.reg.typ.pro.AvroArtifactTypeUtilProvider_Subclass] (executor-thread-3) ENTERING method [getCompatibilityChecker] with 0 parameters
2021-02-27 18:42:11,075 DEBUG [io.api.reg.typ.pro.AvroArtifactTypeUtilProvider_Subclass] (executor-thread-3) LEAVING method [getCompatibilityChecker]
2021-02-27 18:42:11,079 DEBUG [io.sma.met.MetricsRegistryImpl] (executor-thread-3) Register metric [metricId: MetricID{name='rest_http_requests_total', tags=[endpoint="/artifacts/{artifactId}/test",group="REST",method="PUT",metric="rest_http_requests_total",status="4xx"]}, type: counter]
2021-02-27 18:42:11,079 DEBUG [io.sma.met.MetricsRegistryImpl] (executor-thread-3) Register metric [metricId: MetricID{name='rest_http_request_duration', tags=[endpoint="/artifacts/{artifactId}/test",group="REST",method="PUT",metric="rest_http_request_duration",status="4xx"]}, type: timer]
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.core.providerfactory.SortedKey
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.core.providerfactory.SortedKey
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) Interceptor Context: org.jboss.resteasy.core.interception.jaxrs.ServerWriterInterceptorContext,  Method : proceed
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.core.providerfactory.SortedKey
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) MessageBodyWriter: org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider
2021-02-27 18:42:11,080 DEBUG [org.jbo.res.res.i18n] (executor-thread-3) Provider : org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider,  Method : writeTo
2021-02-27 18:42:11,082 INFO  [io.qua.htt.access-log] (executor-thread-3) 127.0.0.1 - - 27/Feb/2021:18:42:11 +0100 "PUT /api/artifacts/user-schema/test HTTP/1.1" 400 12291
```

So something change to latest version on master, you probably more than me here :) It seems to me HTTP 400 indicates misunderstanding between client and server, and not actually that the schema is incompatible after a check? I might be mistaken.

Trying to upload the same file in the UI gives:
```
 Invalid Content Error

The content you attempted to upload violated one or more of the established content rules.

io.apicurio.registry.rules.RuleViolationException: Incompatible artifact: user-schema [AVRO]
	at io.apicurio.registry.rules.compatibility.CompatibilityRuleExecutor.execute(CompatibilityRuleExecutor.java:66)
	at io.apicurio.registry.rules.compatibility.CompatibilityRuleExecutor_Subclass.execute$$superaccessor1(CompatibilityRuleExecutor_Subclass.zig:209)
	at io.apicurio.registry.rules.compatibility.CompatibilityRuleExecutor_Subclass$$function$$1.apply(CompatibilityRuleExecutor_Subclass$$function$$1.zig:33)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.apicurio.registry.logging.LoggingInterceptor.logMethodEntry(LoggingInterceptor.java:55)
	at io.apicurio.registry.logging.LoggingInterceptor_Bean.intercept(LoggingInterceptor_Bean.zig:275)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.apicurio.registry.rules.compatibility.CompatibilityRuleExecutor_Subclass.execute(CompatibilityRuleExecutor_Subclass.zig:157)
	at io.apicurio.registry.rules.compatibility.CompatibilityRuleExecutor_ClientProxy.execute(CompatibilityRuleExecutor_ClientProxy.zig:157)
	at io.apicurio.registry.rules.RulesServiceImpl.applyRule(RulesServiceImpl.java:128)
	...
```

**Comment by EricWittmann** [comment]
Incompatible content should result in a `409`.  I think the only way to get a `400` should be if no content was uploaded.

Hold off on this for now, `master` is going to be massively updated very soon (today maybe) with a lot of changes to the maven mojos.  The information you've provided is very helpful.  Can you attach the two schemas you used to test for compatibility?  (v1 and v2 of the schema)

**Comment by EricWittmann** [comment]
OK we have merged everything into `master` now.  @blofroth if you could attach your test schemas I can give this a try tomorrow.

**Comment by blofroth** [comment]
Here are the schemas:
https://github.com/blofroth/apicurio-registry/tree/reproduce-1248/utils/maven-plugin/src/test/resources/test-build/src/main/resources/schemas

I modified the test-build pom with some profiles, but the commands listed earlier here will reproduce it on my branch.

**Comment by EricWittmann** [comment]
Thanks - I'll use those resources when I get a chance to test this.  Hopefully soon, we're working on 2.0.0.RC1 now.

**Comment by famartinrh** [comment]
@blofroth this test actually verifies the issue you are reporting 
https://github.com/Apicurio/apicurio-registry/blob/master/app/src/test/java/io/apicurio/registry/maven/TestUpdateRegistryMojoTest.java#L52
I think this should be fixed now.
We published 2.0.0.CR1 recently, if you could try it...
`docker run -it -p 8080:8080 apicurio/apicurio-registry-mem:2.0.0.RC1`

**Comment by EricWittmann** [comment]
Marking this as fixed in 2.0 (we think it is).  If this is still a problem, please re-open this issue!
